### PR TITLE
Issue-1488: Quick start guide to developing in OpenMetadata

### PR DESCRIPTION
**Pull request by ksivam**
### Describe your changes :

Improve the developer documentation to include a quick start guide to start developing in OpenMetadata.

#
### Type of change :

- [ ] Bug fix
- [ ] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have performed a self-review of my own. 
- [x] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

@sureshms @harshach

### CODE REVIEW

**Comment by ksivam** [comment]
Note, I will be owning the `quick start guide` to keep up with the changes in the system. Every Monday it will be refreshed for any new changes from the previous week.

File: docs/open-source-community/developer/quick-start-guide.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+### Authentication/Authorization  
+Authentication is provided by google oauth provider. All incoming requests are filtered by validating the JWT token using the google oauth provider. Access control is provided by [`CatalogAuthorizer`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/security/CatalogAuthorizer.java).
+
+Auth/Authz details are configured at [`OpenMetadata/conf/openmetadata-security.yaml`](https://github.com/open-metadata/OpenMetadata/blob/main/conf/openmetadata-security.yaml)
+
+### Ingestion  
+Ingestion is a simple python framework to ingest the metadata from various external sources into OpenMetadata platform.
+
+Ingestion is a simple python framework to ingest the metadata from various external sources into OpenMetadata platform.
```
Comment (mithmatt) [inline docs/open-source-community/developer/quick-start-guide.md:61]: Duplicate line as the one above?

Comment (ksivam) [reply docs/open-source-community/developer/quick-start-guide.md]: thank you. fixed.

**Comment by harshach** [comment]
Thanks @ksivam This is a great start. LGTM.
@shannonbradshaw can you review this as well

**Comment by shannonbradshaw** [comment]
Reviewing now.

File: docs/open-source-community/developer/quick-start-guide.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+## API
+[Dropwizard](https://www.dropwizard.io/) java framework is used to build the RESTful API's.  All the API's are defined under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/resources`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/resources). 
+[Swagger](https://swagger.io/) is used for API documentation following the OpenAPI standards.
+
+
+## System and Components
+![system](https://user-images.githubusercontent.com/3944743/144579545-e51aa37e-93ed-4080-9f32-ce617029571f.png)
+
+### Events  
+Changes to the entities are captured as `events` and are stored in the database and elastic search.
```
[the remark this answers is not in the document]
Comment (shannonbradshaw) [reply docs/open-source-community/developer/quick-start-guide.md:39]: Are change events searchable?

File: docs/open-source-community/developer/quick-start-guide.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+The event handlers are defined under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events) and are applied globally to any outgoing response using the `ContainerResponseFilter`
+
+### Database  
+MySql is used to persist the entities and the repository code to interact with MySql is located under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3).
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:44]:
```suggestion
OpenMetadata uses MySQL for the metadata catalog. The catalog code is located in the directory [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3).
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+The database entity tables are created using the command [`OpenMetadata/bootstrap/bootstrap_storage.sh`](https://github.com/open-metadata/OpenMetadata/blob/main/bootstrap/bootstrap_storage.sh). [Flyway](https://flywaydb.org/) is used for managing the database table versions.
+
+### Elastic Search  
+Entity change events are stored in elastic search. The [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java) is responsible for capturing the change events and updating es.
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:49]: It seems misleading to describe the role of Elasticsearch as solely indexing change events.

File: docs/open-source-community/developer/quick-start-guide.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+Changes to the entities are captured as `events` and are stored in the database and elastic search.
+
+The event handlers are defined under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events) and are applied globally to any outgoing response using the `ContainerResponseFilter`
+
+### Database  
+MySql is used to persist the entities and the repository code to interact with MySql is located under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3).
+
+The database entity tables are created using the command [`OpenMetadata/bootstrap/bootstrap_storage.sh`](https://github.com/open-metadata/OpenMetadata/blob/main/bootstrap/bootstrap_storage.sh). [Flyway](https://flywaydb.org/) is used for managing the database table versions.
+
+### Elastic Search
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:48]:
```suggestion
### Elasticsearch  
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+The database entity tables are created using the command [`OpenMetadata/bootstrap/bootstrap_storage.sh`](https://github.com/open-metadata/OpenMetadata/blob/main/bootstrap/bootstrap_storage.sh). [Flyway](https://flywaydb.org/) is used for managing the database table versions.
+
+### Elastic Search  
+Entity change events are stored in elastic search. The [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java) is responsible for capturing the change events and updating es.
+
+The es indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:51]:
```suggestion
Elasticsearch indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+1. Source → [`OpenMetadata/ingestion/src/metadata/ingestion/source`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/source)
+2. Processor → [`OpenMetadata/ingestion/src/metadata/ingestion/processor`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/processor)
+3. Sink → [`OpenMetadata/ingestion/src/metadata/ingestion/sink`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/sink)
+4. Stage → [`OpenMetadata/ingestion/src/metadata/ingestion/stage`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/stage)
+5. BulkSink → [`OpenMetadata/ingestion/src/metadata/ingestion/bulksink`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/bulksink)
+
+**Airflow**
+
+For simplicity, the metadata from external sources are ingested in a pull-based model, and Apache Airflow is used as an orchestration framework to perform ingestion.
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:84]:
```suggestion
For simplicity, OpenMetadata ingests metadata from external sources using a pull-based model. OpenMetadata uses Apache Airflow to orchestrate ingestion workflows.
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+Example Airflow dags are found under [`OpenMetadata/ingestion/examples/airflow/dags`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/examples/airflow/dags) for reference.
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:86]:
```suggestion
See the directory [`OpenMetadata/ingestion/examples/airflow/dags`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/examples/airflow/dags) for reference DAG definitions.
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+**JsonSchema python typings**
+
+The python typings for OpenMetadata JsonSchema is generated from the [`Makefile`](https://github.com/open-metadata/OpenMetadata/blob/main/Makefile) using the command `make generate`. The generated files are located at `OpenMetadata/ingestion/src/metadata/generated`
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:90]:
```suggestion
You can generate Python types for OpenMetadata models defined using Json Schema using the `make generate` command of the [`Makefile`](https://github.com/open-metadata/OpenMetadata/blob/main/Makefile). Generated files are located in the directory `OpenMetadata/ingestion/src/metadata/generated`
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+### Elastic Search  
+Entity change events are stored in elastic search. The [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java) is responsible for capturing the change events and updating es.
+
+The es indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.
+
+### Authentication/Authorization  
+Authentication is provided by google oauth provider. All incoming requests are filtered by validating the JWT token using the google oauth provider. Access control is provided by [`CatalogAuthorizer`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/security/CatalogAuthorizer.java).
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:54]:
```suggestion
OpenMetadata uses Google OAuth for authentication. All incoming requests are filtered by validating the JWT token using the Google OAuth provider. Access control is provided by [`CatalogAuthorizer`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/security/CatalogAuthorizer.java).
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+Auth/Authz details are configured at [`OpenMetadata/conf/openmetadata-security.yaml`](https://github.com/open-metadata/OpenMetadata/blob/main/conf/openmetadata-security.yaml)
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:56]:
```suggestion
See the configuration file [`OpenMetadata/conf/openmetadata-security.yaml`](https://github.com/open-metadata/OpenMetadata/blob/main/conf/openmetadata-security.yaml) for the authentication and authorization configurations.
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+### Ingestion  
+Ingestion is a simple python framework to ingest the metadata from various external sources into OpenMetadata platform.
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:59]:
```suggestion
Ingestion is a simple Python framework to ingest metadata from external sources into OpenMetadata.
```

File: docs/open-source-community/developer/quick-start-guide.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+**Connectors**
+
+The ingestion is carried out using a set of components called `Connectors` as described [here](https://docs.open-metadata.org/open-source-community/developer/build-a-connector#workflow) and it consists of
```
Comment (shannonbradshaw) [inline docs/open-source-community/developer/quick-start-guide.md:63]:
```suggestion
OpenMetadata defines and uses a set of components called `Connectors` for metadata ingestion. Each data service requires its own connector.  See the documentation on how to [build a connector](https://docs.open-metadata.org/open-source-community/developer/build-a-connector#workflow) for details on developing connectors for new services. 
```

**Review by shannonbradshaw — CHANGES_REQUESTED** [review]

**Comment by shannonbradshaw** [comment]
@ksivam thanks so much for this PR. It's a good start. I know how much effort this type of doc requires. Again, thank you! I've made several requested changes via the "suggestion" tool.

File: docs/open-source-community/developer/quick-start-guide.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
+# Quick start guide
+
+Use this document as a quick start guide to begin developing in OpenMetdata. The document answers `what's` and `where's` of the following topic,
+
+1. Schema, aka models 
+2. API's 
+3. System and Components
+
+## Schema (aka, models)
```
[the remark this answers is not in the document]
Comment (ksivam) [reply docs/open-source-community/developer/quick-start-guide.md:9]: agreed.

File: docs/open-source-community/developer/quick-start-guide.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ksivam) [reply docs/open-source-community/developer/quick-start-guide.md:39]: yes. took your suggestion.

File: docs/open-source-community/developer/quick-start-guide.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+### Entities  
+The entities are located at [`OpenMetadata/catalog-rest-service/src/main/resources/json/schema/entity`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/resources/json/schema/entity), and the following are few examples of the entities supported by OpenMetadata.
+- data
+- feed
+- policies
+- services
+- tags
+- teams
+
+### Common Types
```
[the remark this answers is not in the document]
Comment (ksivam) [reply docs/open-source-community/developer/quick-start-guide.md:23]: sure, changed to "Types".

File: docs/open-source-community/developer/quick-start-guide.md (hunk 4 above)
Comment (ksivam) [reply docs/open-source-community/developer/quick-start-guide.md]: changed the doc, please let me know if this is ok.

OpenMetadata uses Elasticsearch to store the Entity change events and makes it searchable by search index. The [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/elasticsearch/ElasticSearchEventHandler.java`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/elasticsearch/ElasticSearchEventHandler.java) is responsible for capturing the change events and updating Elasticsearch.

Elasticsearch indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.

**Comment by ksivam** [comment]
> @ksivam thanks so much for this PR. It's a good start. I know how much effort this type of doc requires. Again, thank you! I've made several requested changes via the "suggestion" tool.

@shannonbradshaw thank you very much for the detailed review. Really appreciate your time to make this PR better.

Resolved and accepted all your suggestions. Please let me know if there are any more changes.

**Review by shannonbradshaw — APPROVED** [review]

**Comment by shannonbradshaw** [comment]
Thanks, @ksivam . @harshach I've approved this PR.

**Comment by shannonbradshaw** [comment]
@harshach do we need to cherry pick this to 0.6.0 so that it appears in the docs? I'm not seeing it.

Maybe you're waiting to make it available in the 0.7.0 docs?

**Pull request merged**